### PR TITLE
[SFT Trainer] precompute packed iterable into a dataset

### LINKED ISSUE #1008: SFTTrainer training stops early?
It seems it just happen today. Has anyone met the error where the training stops early?  For example, it should have run 1000 steps but stops after 300 steps, and giving a output like:

`TrainOutput(global_step=100, training_loss=1.2991660499572755, metrics={'train_runtime': 855.9696, 'train_samples_per_second': 1.869, 'train_steps_per_second': 0.117, 'total_flos': 6.50352940548096e+16, 'train_loss': 1.2991660499572755, 'epoch': 0.16})`

I did my training on Colab, and i don't know if it is an internet issue. But my previous training never meets this kind of issue.

### PR DESCRIPTION
**Pull request by lvwerra**
With this PR we'll precompute the packed iterable dataloader and push it into a dataset. That will make sure the epoch estimate is right and we have a working progress bar. 

Fixes #1008 #1004

### CODE REVIEW

**Review by lewtun — APPROVED** [review]
Great fix @lvwerra - very clean and elegant 🔥 !

Have you run the SFT example script to verify the training steps are correct & learning rate scheduler etc all look OK?

**Review by younesbelkada — APPROVED** [review]
Looking great, thanks @lvwerra !
I tested this PR with the script below:
```python
from transformers import AutoModelForCausalLM, TrainingArguments
from datasets import load_dataset
from trl import SFTTrainer

# This should get ~700 examples
dataset = load_dataset("imdb", split="train[:5%]")

model = AutoModelForCausalLM.from_pretrained("facebook/opt-125m")

args = TrainingArguments(
    output_dir="./",
    per_device_train_batch_size=2,
    num_train_epochs=1,
    logging_steps=1,
    lr_scheduler_type="cosine",
)

trainer = SFTTrainer(
    model,
    args=args,
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=512,
    packing=True,
)

trainer.train()
```
And it gives me correct LR curves:
Screenshot 2023-12-04 at 13 02 07

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -376,6 +369,52 @@ def tokenize(element):
...
+                num_of_sequences=num_of_sequences,
+                chars_per_token=chars_per_token,
+                eos_token_id=tokenizer.eos_token_id,
+            )
+
+            def data_generator(constant_length_iterator):
+                for i in constant_length_iterator:
+                    yield i
+
+            try:
```
[the remark this answers is not in the document]
Comment (chujiezheng) [reply trl/trainer/sft_trainer.py:402]: Same question here
